truffle/src/main/java/org/jruby/truffle/parser/BodyTranslator.java

```diff
@@ -1826,14 +1826,17 @@ public RubyNode visitHashNode(HashParseNode node) {
         final SourceSection fullSourceSection = sourceSection.toSourceSection(source);
 
         final List<RubyNode> hashConcats = new ArrayList<>();
+        final List<Boolean> hashConcatsIsMerge = new ArrayList<>();
 
         final List<RubyNode> keyValues = new ArrayList<>();
 
         for (KeyValuePair<ParseNode, ParseNode> pair: node.getPairs()) {
             if (pair.getKey() == null) {
                 final RubyNode hashLiteralSoFar = HashLiteralNode.create(context, fullSourceSection, keyValues.toArray(new RubyNode[keyValues.size()]));
                 hashConcats.add(hashLiteralSoFar);
+                hashConcatsIsMerge.add(false);
                 hashConcats.add(HashCastNodeGen.create(context, fullSourceSection, pair.getValue().accept(this)));
+                hashConcatsIsMerge.add(true);
```

Comment:
Why such a complex approach rather than introducing a node above the HashCastNode here that checks keys are all Symbols like `EnsureSymbolKeysNode` ?
Then that would be a single line change here, and no change in ConcatHashLiteralNode.
Don't be afraid to add new nodes.